nixos/modules/services/matrix/conduwuit.nix

```diff
@@ -0,0 +1,299 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.conduwuit;
+  defaultUser = "conduwuit";
+  defaultGroup = "conduwuit";
+
+  format = pkgs.formats.toml { };
+  # TOML does not allow null values, so we just omit those fields
+  filteredSettings = lib.converge (lib.filterAttrsRecursive (_: v: v != null)) cfg.settings;
+  configFile = format.generate "conduwuit.toml" filteredSettings;
+in
+{
+  meta.maintainers = with lib.maintainers; [ niklaskorz ];
+  options.services.conduwuit = {
+    enable = lib.mkEnableOption "conduwuit";
+
+    user = lib.mkOption {
+      type = lib.types.nonEmptyStr;
+      description = ''
+        The user {command}`conduwuit` is run as.
+      '';
+      default = defaultUser;
+    };
+
+    group = lib.mkOption {
+      type = lib.types.nonEmptyStr;
+      description = ''
+        The group {command}`conduwuit` is run as.
+      '';
+      default = defaultGroup;
+    };
+
+    extraEnvironment = lib.mkOption {
+      type = lib.types.attrsOf lib.types.str;
+      description = "Extra Environment variables to pass to the conduwuit server.";
+      default = { };
+      example = {
+        RUST_BACKTRACE = "yes";
+      };
+    };
+
+    package = lib.mkPackageOption pkgs "conduwuit" { };
+
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = format.type;
+        options = {
+          global.server_name = lib.mkOption {
+            type = lib.types.nonEmptyStr;
+            default = "";
+            example = "example.com";
+            description = "The server_name is the name of this server. It is used as a suffix for user # and room ids.";
+          };
+          global.address = lib.mkOption {
+            type = lib.types.nullOr (
+              lib.types.oneOf [
+                lib.types.nonEmptyStr
```

Comment:
Why noneEmptryStr here? null or list isn't enough?

Reply:
I was following the config types conduwuit accepts, which is either a string literal or a list of strings, or not specifying at all (in our case, `null`). I'm fine with going with list-only if @girlbossceo has no objections.

Reply:
@niklaskorz is correct; these are perfectly valid usages of the `address` and `port` config options:

```
# listen only on IPv4 localhost
address = "127.0.0.1"

# listen only on IPv4 and IPv6 localhost (the default)
address = ["127.0.0.1", "::1"]

# this is also perfectly valid to listen on only IPv4 localhost
address = ["127.0.0.1"]

(the same applies for port)
```

When setting a UNIX socket, this **must** be unset. If `null` in Nix translates to that, then it must be that.

So from my understanding, this should be left as is.

Reply:
Is not setting the option not equal to setting the option to emptry string? Should we then not convert null options to an empty option instead?

Reply:
An empty string is not interpreted the same. Conduit users have tried this when they used the Conduit module on top of conduwuit in an attempt to make UNIX sockets work, and it does not work. The underlying type in Rust for these options is an `IpAddr`, an empty string does not parse to that and will fail config deserialisation. It's unnecessary complexity and potential mass config breakage if I change these to strings and start converting them. 

It's the same situation with the `port` config option, the underlying type is a `u16`. An empty string does not parse to that and will fail config deserialisation.

I can't comment on the converting null part, but there is a clear difference between unsetting a config option by not specifying it, and setting it to an empty string. And I can say that for UNIX sockets to work, it needs to be unset completely, not set to an empty string or a null value.

I have posted an example already at the start of the PR of how this config module should translate to TOML for UNIX sockets to work.

Reply:
As toml does not have a concept of `null`, it's fine to translate it to leaving out the value. As for the complex types, I've changed address and port to accept a list and not "either a list or a single value".